lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp

```diff
@@ -1520,258 +1719,98 @@ LogicalResult FIRRTLLowering::visitDecl(MemOp op) {
-      SmallVector<WritePipeElement> writePipe;
-      auto rdEn = getPortFieldValue("en");
-      auto rdAddr = getPortFieldValue("addr");
-      auto rdMask = getPortFieldValue("mask");
-      auto rdData = getPortFieldValue("data");
-      writePipe.push_back({portWires["en"], portWires["addr"],
-                           portWires["mask"], portWires["data"], rdEn, rdAddr,
-                           rdMask, rdData});
-
-      // Construct wripe pipe registers for non-unary write latency
-      Value wRegEn, wRegAddr, wRegMask, wRegData;
-      for (size_t j = 0; j < writeLatency - 1; ++j) {
-        if (j == 0) {
-          wRegEn = createArrayReg("_en_pipe", i1Type, writeLatency - 1);
-          wRegAddr = createArrayReg("_addr_pipe", addrType, writeLatency - 1);
-          wRegMask = createArrayReg("_mask_pipe", i1Type, writeLatency - 1);
-          wRegData = createArrayReg("_data_pipe", dataType, writeLatency - 1);
-        }
-        auto jIdx = getOrCreateIntConstant(log2(writeLatency), j);
-        auto arEn = builder.create<sv::ArrayIndexInOutOp>(wRegEn, jIdx);
-        auto arAddr = builder.create<sv::ArrayIndexInOutOp>(wRegAddr, jIdx);
-        auto arMask = builder.create<sv::ArrayIndexInOutOp>(wRegMask, jIdx);
-        auto arData = builder.create<sv::ArrayIndexInOutOp>(wRegData, jIdx);
-        writePipe.push_back({arEn, arAddr, arMask, arData,
-                             builder.create<sv::ReadInOutOp>(arEn),
-                             builder.create<sv::ReadInOutOp>(arAddr),
-                             builder.create<sv::ReadInOutOp>(arMask),
-                             builder.create<sv::ReadInOutOp>(arData)});
-      }
-
-      // Build the write pipe for non-unary write latency
-      if (writeLatency != 1) {
-        addToAlwaysFFBlock(sv::EventControl::AtPosEdge, clk, [&]() {
-          for (size_t j = 1; j < writeLatency; ++j) {
-            buildPAssign(writePipe[j].en, writePipe[j - 1].rd_en);
-            addIfProceduralBlock(writePipe[j - 1].rd_en, [&]() {
-              buildPAssign(writePipe[j].addr, writePipe[j - 1].rd_addr);
-              buildPAssign(writePipe[j].mask, writePipe[j - 1].rd_mask);
-              buildPAssign(writePipe[j].data, writePipe[j - 1].rd_data);
-            });
-          }
-        });
-      }
-
-      // Attach the write port
-      auto last = writePipe.back();
-      auto enable = last.rd_en;
-      auto cond = builder.createOrFold<comb::AndOp>(enable, last.rd_mask);
-      auto slot = builder.create<sv::ArrayIndexInOutOp>(reg, last.rd_addr);
-      auto rd_lastdata = last.rd_data;
-      addToAlwaysFFBlock(sv::EventControl::AtPosEdge, clk, [&]() {
-        addIfProceduralBlock(cond, [&]() { buildPAssign(slot, rd_lastdata); });
-      });
-      continue;
-    }
+    if (portKind == MemOp::PortKind::Read) {
+      addInput(readOperands, "clk", 1);
+      addInput(readOperands, "en", 1);
+      addInput(readOperands, "addr", llvm::Log2_64_Ceil(memSummary.depth));
```

Comment:
I think these uses of Log2_64_Ceil will also need the `std::max(1, ...)` treatment.

```suggestion
      addInput(readOperands, "addr", std::max(1U, llvm::Log2_64_Ceil(memSummary.depth)));
```

Can probably just compute that once before the if/else branch

Reply:
The max was moved into add Input and should be removed there, thanks.

Reply:
Actually, I was testing with the latest revision, but there still appears to be an issue. The max is used here: https://github.com/llvm/circt/pull/864/files#diff-fb69b8a7ca0d098b8278662283e61f2dd1564898caa902d72c79b2299f833a8eR1748, but below the `if (width)` doesn't take into account the fact that width may have been bumped from 0 to 1. Is that intended? LowerToRTL was failing with an assert the way it is now, for a depth-1 memory.